packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -440,11 +441,105 @@ class RenderParagraph extends RenderBox {
+    text.visitTextSpan((TextSpan span) {
+      if (span.recognizer != null) {
+        if (buffer != null) {
+          final SemanticsNode child = new SemanticsNode();
+          final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+          child.rect = _getRectForSpan(start, end);
+          childConfig
+            ..label = buffer.toString()
+            ..textDirection = textDirection
+            ..sortKey = new OrdinalSortKey(order++);
+          buffer = null;
+          child.updateWith(config: childConfig);
+          newChildren.add(child);
+          start = end;
+        }
+        end += span.text.length;
+        final SemanticsNode child = new SemanticsNode();
+        final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+        childConfig.textDirection = textDirection;
+        childConfig.label = span.text;
+        childConfig.sortKey = new OrdinalSortKey(order++);
+        if (span.recognizer is LongPressGestureRecognizer) {
+          final LongPressGestureRecognizer recognizer = span.recognizer;
+          childConfig.onLongPress = recognizer.onLongPress;
+        } else if (span.recognizer is TapGestureRecognizer) {
+          final TapGestureRecognizer recognizer = span.recognizer;
+          childConfig.onTap = recognizer.onTap;
+        }
+        child.rect = _getRectForSpan(start, end);
+        child.updateWith(config: childConfig);
+        newChildren.add(child);
+        start = end;
+      } else {
+        buffer ??= new StringBuffer();
+        buffer.write(span.text);
+        end += span.text.length;
+      }
+      return true;
+    });
+    if (buffer != null && buffer.isNotEmpty) {
+      final SemanticsNode child = new SemanticsNode();
+      final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+      child.rect = _getRectForSpan(start, end);
+      childConfig
+        ..sortKey = new OrdinalSortKey(order++)
+        ..label = buffer.toString()
+        ..textDirection = textDirection;
+      child.updateWith(config: childConfig);
+      newChildren.add(child);
+    }
+    node.updateWith(config: config, childrenInInversePaintOrder: newChildren);
+  }
+
+  Rect _getRectForSpan(int start, int end) {
+    final TextSelection current = new TextSelection(baseOffset: start, extentOffset: end);
+    final List<ui.TextBox> currentBoxes = getBoxesForSelection(current);
+    assert(currentBoxes.isNotEmpty);
+    final double top = currentBoxes.first.top;
+    double left = currentBoxes.first.left;
+    double right = currentBoxes.first.right;
+    double bottom = currentBoxes.first.bottom;
+    for (TextBox box in currentBoxes.skip(1)) {
+      print('$box');
+      if (box.left < left)
+        left = box.left;
+      if (box.right > right)
+        right = box.right;
+      if (box.bottom > bottom)
+        bottom = box.bottom;
+    }
+    return new Rect.fromLTRB(left, top, right, bottom);
```

Comment:
Rect has more idiomatic ways of doing this, e.g. Rect.expandToInclude

Reply:
e.g. you could just do:
```dart
return getBoxesForSelection(current).reduce((Rect result, Rect box) => result.expandToInclude(box));
```

Reply:
Ahh reduce...I always forget about it for `fold`, which doesn't work in this case.

Reply:
Since I am tracking the directionality as well I still need to loop, but now I use expandToInclude